src/core/impl/TChem_Impl_MESA.hpp

```diff
@@ -0,0 +1,1891 @@
+    const ordinal_type jliquid=2;
+    const ordinal_type jtotal=3;
+
+    const ordinal_type no_aerosol=0;
+    const ordinal_type all_solid=1;
+    const ordinal_type all_liquid=2;
+    const ordinal_type mixed=3;
+
+    const ordinal_type soluble=1;
+    const ordinal_type insoluble=2;
+
+    const ordinal_type jhyst_up=1;
+    const ordinal_type jhyst_lo=0;
+
+    // polynomial coefficients for binary molality (used in ZSR)
+       // for aw < 0.97
+    real_type_2d_dual_view a_zsr; // ("a_zsr", 6, nelectrolyte);
+    real_type_1d_dual_view aw_min; // ("aw_min", nelectrolyte);
+    real_type_1d_dual_view b_zsr; // ("b_zsr", nelectrolyte)
+
+    // local aerosol ions
+    // cations
+    const ordinal_type jc_h   = 0;
+    const ordinal_type jc_nh4 = 1;
+    const ordinal_type jc_na  = 2;
+    const ordinal_type jc_ca  = 3;
+
+    // anions
+    const ordinal_type ja_hso4 = 0;
+    const ordinal_type ja_so4  = 1;
+    const ordinal_type ja_no3  = 2;
+    const ordinal_type ja_cl   = 3;
+    const ordinal_type ja_msa  = 4;
+    // const ordinal_type ja_co3  = 6; (note: this is commented out in MOSAIC as well)
+
+    // MESA parameters
+    const ordinal_type Nmax_MESA = 80;
+    const real_type rtol_mesa = 0.01;
+
+    // Temperature-dependent thermodynamic parameters
+    // liquid-liquid
+    // const ordinal_type nrxn_aer_ll = 3;
+    real_type_1d_dual_view Keq_298_ll; // ("Keq_298_ll", nrxn_aer_ll);
+    real_type_1d_dual_view Keq_a_ll; // ("Keq_a_ll", nrxn_aer_ll);
+    real_type_1d_dual_view Keq_b_ll; // ("Keq_b_ll", nrxn_aer_ll);
+
+    // parameters for MTEM mixing rule (Zaveri, Easter, and Wexler, 2005)
+    real_type_3d_dual_view b_mtem; // ("b_mtem", 6, nelectrolyte, nelectrolyte);
+
+    // molecular weights for aerosol species
+    real_type_1d_dual_view mw_aer_mac;
+
+    // densities for aerosol species
+    real_type_1d_dual_view dens_aer_mac;
+
+    // molecular weights for electrolytes
+    real_type_1d_dual_view mw_electrolyte;
+
+    // magnitude of charges on ions
+    ordinal_type_1d_dual_view zc;
+    ordinal_type_1d_dual_view za;
+
+    // molecular weights of ions [g/mol]
+    real_type_1d_dual_view MW_c;
+    real_type_1d_dual_view MW_a;
+};
+
+template<typename DeviceType>
+struct MOSAIC_ConstData {
+public:
+  using device_type = DeviceType;
+
+  using real_type_1d_view_type = Tines::value_type_1d_view<real_type, device_type>;
+  using real_type_2d_view_type = Tines::value_type_2d_view<real_type, device_type>;
+  using real_type_3d_view_type = Tines::value_type_3d_view<real_type, device_type>;
+
+  using ordinal_type_1d_view_type = Tines::value_type_1d_view<ordinal_type, device_type>;
+
+  using mcd_real_type_1d_view = ConstUnmanaged<real_type_1d_view_type>;
+  using mcd_real_type_2d_view = ConstUnmanaged<real_type_2d_view_type>;
+  using mcd_real_type_3d_view = ConstUnmanaged<real_type_3d_view_type>;
+
+  using mcd_ordinal_type_1d_view = ConstUnmanaged<ordinal_type_1d_view_type>;
+
+  mcd_real_type_2d_view a_zsr;
```

Comment:
We must allocate these views and can create a dedicated routine for this purpose.